### DISCUSSION TITLE
DualSense support

### DISCUSSION THREAD

**Discussion by loveisfoss** [v1993/evdevhook#11] [General]
Hey, thanks for this awesome piece of work!

What does the following mean?
`hid-playstation - DualSense controller (Note: not in mainline kernel yet as of now)`

Using Debian bookworm with kernel 5.16. The DualSense controller is working without problems, generally. I suppose DualSense support has been included in kernel 5.16, is this right? But when trying to use evdevhook:

```
$ ./evdevhook
Connected motion devices:
$
```
...shows no motion devices.

`$ ./evdevhook ./config_templates/playstation.json
 `
...stays here forever and does nothing.

After a little web searching I wondered if I should modprobe the playstation module (although the controller is working fine already), but I got this:

```
$ sudo modprobe hid-playstation
modprobe: FATAL: Module hid-playstation not found in directory /lib/modules/5.16.0-1-amd64
```

Any ideas?

**Comment by v1993** [comment]
I'm not quite sure about current status of dualsense support in linux kernel - I don't own dualsense myself (nor plan to get one). When the quoted sentence was being written a separate DKMS module still was requires for DS5 support - that's what I know.

You can try running `sudo chmod a+rw /dev/input/event*` and checking if motion device appears after that - if it does, you need to set up udev rule for access, if it does not - you need to install DKMS module. Sadly, I don't really have time to help in depth with either outcome - you'll have to figure those out yourself.

Finally, check out #3 and edit mapping in `./config_templates/playstation.json` to what is suggested there (it also contains an explanation why I haven't added this manually... yet) to have motion work correctly.

**Comment by loveisfoss** [comment]
Thanks for your reply! Here come some observations.

The driver hid-playstation seems to be present on kernel 5.15 and not on kernel 5.16 in my Debian installation. A notable difference is that the moving of cursor by touchpad can be used when booting with kernel 5.15 but not with 5.16. 

On kernel 5.16
The controller seemed to work without problems. Even native motion controls of DualSense work on yuzu (the native one, named sdl - not with DSU). So, it seems that something has possibly changed on kernel 5.16 and the module hid-playstation is not used anymore, but the controller is working without problems (minus the touchpad cursor control). It is not being detected by evdevhook, though.

On kernel 5.15:
The module hid-playstation exists and is being loaded AFTER the controller is being paired/connected to the pc. If one does `lsmod | grep -i playstation` without having the controller connected, it is not going to show anything. But the module is being loaded normally after the controller is connected and you can see it by doing lsmod.

The command you mention is needed in my case.
`sudo chmod a+rw /dev/input/event*`
After doing this, evdevhook detects the motion device of DualSense (but only on kernel 5.15 where the module hid-playstation exists).

Tested on yuzu and it works (meaning not the native one, named sdl - but through DSU).

Tested on cemu/cemuhook (running with wine/lutris) and it seems to be detected in the cemuhook submenu. It is not working in-game though and the option to use motion controls ingame is grayed out. This may have to do with my lutris/wine implementation, I may look into it later. @Mrbinaer how did you make it work on cemu? Do you run cemu with wine?

**Reply by Mrbinaer** [reply]
Yes, I am running cemu via the wine version lutris-5.7 and kernel version 5.15. After the cemu update you can add the evdevhook motion source via the DSUController API and enabling "Use Motion" under settings.

**Comment by v1993** [comment]
On cemu you must use gamepad for gyro to work. I'd also recomment against using cemuhook, especially that now cemu supports software video decoding, removing the last issue cemuhook solved.

No idea what's the deal with kernel module status, sadly.

**Comment by loveisfoss** [comment]
My idea is to run a DSU client on cemu, in this case cemuhook. Is this no longer needed for motion controls on cemu?

**Reply by v1993** [reply]
Cemu now supports DSU motion without cemuhook (remove cemuhook if you use it - it loves to get in the way).

**Comment by loveisfoss** [comment]
So, I tried to uninstall and reinstall the 5.16 kernel and I see now an error when installing it.

`Error! Bad return status for module build on kernel: 5.16.0-1-amd64 (x86_64)`

It seems like what I wrote earlier about kernel 5.16 being different is wrong. So sorry X.X.
We can possibly end this discussion here. Let's see how things develop on their own about this driver.

**Reply by v1993** [reply]
You probably need to update dkms module (which you no doubt have installed for dualsense) - there's a decent chance it was updated for 5.16 by now.

**Comment by loveisfoss** [comment]
udev rule I use for DualSense:
Create a file: /etc/udev/rules.d/50-ds-motion.rules
`ATTRS{name}=="Wireless Controller Motion Sensors", OWNER="username", MODE="0660"`
evdevhook can now connect without the need for extra chmod command.

It is even working on cemu after a fresh reinstall via lutris. But not being recognised under option "DSU controller" in input settings. I don't know why, as this option works for @Mrbinaer .

It works only when I set the controller as a DirectInput one and I use cemuhook as a DSU client that gets info from evdevhook.
Dualsense mapping change was also needed, as mentioned in #3 .

So, cemu works now with motion controls and DualSense on Linux. God job! :)

ps. Kernel 5.15 with hid-playstation existing by default.

**Comment by loveisfoss** [comment]
So, it seems like the module hid-playstation is really missing from debian kernel 5.16 (at least for now).

https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1844706.html

It seems like the module hid_generic is being used instead, with which button mapping is kind of wrong and gyro/motion is not working.
Just thought that I would leave this here for anyone who may be struggling with this.

**Comment by SuperSamus** [comment]
For those who were wondering, hid-playstation was merged in Linux 5.12.
